emcc.py

```diff
@@ -625,6 +625,13 @@ def check_human_readable_list(items):
       passes += ['--pass-arg=asyncify-onlylist@%s' % ','.join(shared.Settings.ASYNCIFY_ONLY)]
   if shared.Settings.BINARYEN_IGNORE_IMPLICIT_TRAPS:
     passes += ['--ignore-implicit-traps']
+  # normally we can assume the memory, if imported, has not been modified
+  # beforehand (in fact, in most cases the memory is not even imported anyhow,
+  # but it is still safe to pass the flag). the one exception is dynamic
+  # linking of a side module: the main module is ok as it is loaded first, but
+  # the side module may be assigned memory that was previously used.
+  if run_binaryen_optimizer and not shared.Settings.SIDE_MODULE:
+    passes += ['--unmodified-imported-mem']
```

Comment:
Does this flag basically mean a clean, zero-initialized memory is assumed?

(If so I think maybe we could have come up with better name)

Reply:
Yes, it's not modified, so it begins as all zeros.

Happy to do a better name. What do you suggest?

Reply:
I'm not sure I can think of a better name... but that fact that is wasn't immediately obvious to me makes me think we can do better.

It really the zero'd state, no the modified bit that we care about.  I think it basically mean `--assume-static-data-region-pre-zero'd`.   For me personally I might even prefer something like `--skip-bss` which I imagine refers to precisely what this flag is doing.   Or `--no-zero-bss`?    

BTW, without this flag shouldn't we really zero the entire heap too.. not just the bss regio?  Or maybe sbrk isn't defined to return zero initialized pages?

Reply:
According to the `sbrk` wikipedia entry: `The available space is initialized to a value of zero, unless the break is lowered and then increased`.

Reply:
The heap is automatically zeroed by wasm, except for the active segments in the wasm, but maybe I misunderstood what you were saying?

I don't have a strong preference between the names. But "unmodified" make a little more sense in my head since what we are guarding against is the memory being *used* by something, before the wasm sees it. For example, if we re-enable the ctor evaller, it would mean we stop passing this flag.

Reply:
FWIW Node.js called its option `--zero-fill-buffers` when they added ability to create zero-filled Buffers. Maybe `--assume-zero-filled-memory`?

Reply:
Yes, zero-filled I believe is more meaningful I believe.  I could modify a memory .. then memset it back to zero for the next program.. now it *is* zero filled by still modified.  Thats why I think the modified adjective doesn't quite fit.   But this is really an internal detail and not very important.

Reply:
> The heap is automatically zeroed by wasm, except for the active segments in the wasm, but maybe I misunderstood what you were saying?

But not if the memory is re-used between programs, or between runs of a program.

IIUC if I don't pass `--unmodified-imported-mem` to binaryen that I have to assume that all of memory is garbage, including the heap, right?

Reply:
I guess what I'm saying is that we don't handle dirty memory in general.. only dirty memory in the BSS section (since we have a way to zero'ing that in case it is dirty.. we don't currently have any way of zeroing the heap or the stack).

This is why think the flag name should probably involve the terms `bss` or `static data region` or something like that.   But again, not a big deal.

Reply:
I think `--assume-zero-filled-memory` would be a good option.    But in the absence of this flag we should probably memset the stack and heap too if we want to be honest about it.

Reply:
This PR has been updated to use the new name.